### ISSUE TITLE
Log DMR messages/SMS to text file

### ISSUE THREAD

**Issue by alphapats** [lwvmobile/dsd-fme#186]
Presently, dsd-fme displays received DMR messages and other metadata of all decoded calls on terminal. Requested that following options are added:
1. All intercepted messages/SMS/ Latlong etc of intercepted links are recorded into text file.
2. Complete metadata of decoded links/calls is stored in log file.

Simple option is to  pipe the complete dsd output to text file, but I was looking for more cleaner and structured approach so that  subsequently intercepted metadata can be imported to database (SQLite) for more analysis.

**Comment by kahara** [comment]
I'd definitely find something like this useful, too, with the exception that I'd like to have **all** the decoded data available, preferably as [JSON lines](https://jsonlines.org/). Not sure I understand what the difference between _"into text file"_ and _"in log file"_ in 1. and 2. is(?).

But would it make sense to do it just for DMR? Or put another way, if someone starts implementing such a feature, should it maybe be built in such a way that other modes could easily be plugged into this "advanced logger" later on?

Also, designing the logging format in such a way that it can be extended later or may not/should not be considered a trivial task.

I should be able to do this, but would likely require a little guidance to get started.

**Comment by kahara** [comment]
@lwvmobile what do you think, would this kind of more-structured-logging make sense?

**Comment by lwvmobile** [comment]
> @lwvmobile what do you think, would this kind of more-structured-logging make sense?

If its something you want to work on, then by all means, go ahead. Other than that, I really can't suggest a good starting point for your project, since data logging is spread out all over the place. I would bargain that doing such work is not going to be a trivial task however, really just depends on how deep you want to go with it. If you work on it and make commits to your own fork, when it get to a maturity level, then I may consider pulling it into a development branch on this end, but for the time being, I don't have any plans to do any heavy coding or substantial rewrites to add new features, like a logging system. The only work on DSD-FME I'll be doing in the foreseeable future is just tweaks to trunking and minor fixes as required. I'm not currently taking on any new feature requests.

**Comment by alphapats** [comment]
@lwvmobile Atleast option to write message output to text file can be added?

**Comment by lwvmobile** [comment]
> Atleast option to write message output to text file can be added?

LRRP messages and other location messages already are, when known, but I don't have any handling for most text messages, mostly because its hard to get actual samples with known text messages in them. I do have handling for TIII UDT text messages as is in the DMR ETSI manual, but Moto and other vendor specific versions, I don't have enough information on how they are handled to know when the message is a text vs any other arbitrary information sent out in data blocks.

**Comment by romanremus** [comment]
LWVMOBILE, these messages are transmitted by a DMR channel, I started DSDPlus in parallel with DSD-FME, but in DSDFME it does not display these messages, do you think you will be able to insert them in the future, for their decoding?


![LWVMOBILE](https://github.com/lwvmobile/dsd-fme/assets/54954927/a6174859-02a5-456b-a29b-ae8d15c40f61)

**Comment by lwvmobile** [comment]
@romanremus Can you send some wav file samples of that when that occurs?

**Comment by romanremus** [comment]
of course, as soon as I manage to register, I will gladly send you LWVMOBILE

**Comment by romanremus** [comment]
Hello Lwvmobile, just a question, could you make sure we have a terminal that works in Windows? , I tried something with PDCurses but I have no good results, or could you create DSD-FME with an interface like DSDPlus? the code is super good, but in order to be able to enjoy the call alert or to see certain data, it must be run only in cygwin, or at least you can help me to create an interface, thank you very much

**Comment by TeddJohnson** [comment]
> Hello Lwvmobile, just a question, could you make sure we have a terminal that works in Windows? , I tried something with PDCurses but I have no good results, or could you create DSD-FME with an interface like DSDPlus? the code is super good, but in order to be able to enjoy the call alert or to see certain data, it must be run only in cygwin, or at least you can help me to create an interface, thank you very much

@romanremus I am able to run it just fine in powershell, maybe open a new ticket to discuss that issue?

**Comment by romanremus** [comment]
@teddjohnson  Hello, Thank you for the answer, I really liked that DSD-FME could have an interface that could work in Windows, like DSDPlus for example, I tried some options, I'm not a programmer, do you think it can be created?

**Comment by lwvmobile** [comment]
I might be willing to import the event logger from [M17-FME](https://github.com/lwvmobile/m17-fme) and work on it over time at this point as I have a firmer (but not complete) grasp on DMR PDUs, among other things. Keep in mind though, if I do work on this, it will be a slow and gradual process, with only certain things getting into the event log at the beginning.

Depending on how much of a time investment would be needed to do this, I may work on converting the entire 'call history' to work akin to how M17-FME does, since when I started writing that last year, I kept this issue in the back of my mind on how to tie these things in together, but the difficult issue would be tracking multiple system types, as opposed to a single much simpler system that is well documented and open source.

**Comment by romanremus** [comment]
hello everyone......I think that if lwvmobile created a console like in dsd plus it would be the best decoding software I know, I also tried different options to create a console with menus, like dsd plus, but I only got errors, programming is not for everyone

**Comment by romanremus** [comment]
Hello Lvwmobile, now with the new update from you in Ncurses, on a DMR signal with RSP REQ, it always appears at IP Based SRC - 0, is it a problem or is it ok like this?

**Comment by lwvmobile** [comment]
I'm not sure if that's a problem or not, source shouldn't be zero, unless there is some error in the header. I've not observed that issue on any of the systems I monitor, and that string in ncurses blips by so fast I can't even see it most times, it just changes over to whatever the PDU contents are (LRRP, TMS, etc). Actually, I take that back, I just caught one on my end, but it had a source and target populated. 

Any particular system type on that? Cap+, XPT, anything like that, or just MotoTRBO DMR?

**Comment by romanremus** [comment]
Hello LWVMOBILE, it's just a MOTO TRBO, I'm trying to attach a picture so you can see better.

**Comment by romanremus** [comment]
![Image](https://github.com/user-attachments/assets/d169da5d-0c59-41a6-982f-82068e6ba83e)

**Comment by romanremus** [comment]
TGT always this

**Comment by lwvmobile** [comment]
Without a capture bin or some samples to playback when that occurs, I wouldn't be able to tell you what's going on with it.

**Comment by romanremus** [comment]
https://we.tl/t-3P6Jp9Aa7C

**Comment by romanremus** [comment]
I have attached a RAW recording from the Captured DMR signal, Thank you LWVMOBILE

**Comment by lwvmobile** [comment]
> I have attached a RAW recording from the Captured DMR signal, Thank you LWVMOBILE

I made a new issue referencing what is occurring in your samples, so please continue over at

https://github.com/lwvmobile/dsd-fme/issues/282

**Comment by lwvmobile** [comment]
> Presently, dsd-fme displays received DMR messages and other metadata of all decoded calls on terminal. Requested that following options are added:
> 
>     1. All intercepted messages/SMS/ Latlong etc of intercepted links are recorded into text file.
> 
>     2. Complete metadata of decoded links/calls is stored in log file.
> 
> 
> Simple option is to pipe the complete dsd output to text file, but I was looking for more cleaner and structured approach so that subsequently intercepted metadata can be imported to database (SQLite) for more analysis.

Alright, this is out now, use `-J event_log.txt` as your option, and it'll dump all this stuff to an event log for you.

With that, I'm closing this as completed.
